### DISCUSSION TITLE
Improvements to be made to results screen score, and arguments for why it should be the path forward

### DISCUSSION THREAD

**Discussion by Rekunan** [ppy/osu#23947] [Ideas]
ScoreV2 was recently implemented as the current lazer scoring algorithm, and it's on its way to becoming the permanent scoring algorithm. However, I'd like to make a case for why I disagree with the direction of unrecalculateable scores in osu!lazer, and improvements that can be made to the results screen based system that was in place.

(co-authored by @Natelytle)

To start with the reasons:

- There is a planned time period where people will provide feedback and suggest improvements. However, if lazer score changes according to this, the current conversion method will only get less accurate. The only reason ScoreV1 can be converted to ScoreV2 so accurately at the moment is because the combo portion of ScoreV2 is very similar to ScoreV1. However, if for any reason the combo scaling formula is changed for balancing reasons, this accuracy would decrease, potentially to a level undesirable for leaderboards.

- There are ScoreV1 edge cases that need accounting for. As an example, the bottom diff of [this mapset](https://osu.ppy.sh/beatmapsets/325158#osu/741477) has a multiplier of 0 on the combo scaling portion, which means every judgement is worth the same amount. This causes a linear score progression curve, as opposed to a quadratic one. This is a problem, since if the score is converted using the proposed methods, scores with the same combo value set in lazer will be way less than their corresponding stable scores. This is the only occurrence of this problem, but it's still important to consider.

- People like Zyf applaud the new direction for scoring using more information than is available on the results screen, for reasons I disagree with. He has the standpoint that score should provide information not on the results screen, such as punishing spread out misses more than grouped misses, rewarding longer non-max combos, etc. This is all completely subjective and is not how I view score. Despite having these features, people dislike ScoreV1 for putting plays with worse max combo, acc, misscount, etc., above plays with better values in those metrics. You can say that this is because ScoreV1 goes about it wrong, but I think there is no way to do this "correctly", and the reason is the fact that you are not watching 2 plays live side by side in a tournament environment, but instead using the end results screen.

- Similar to the previous point, it is easy to get very close to approximating ScoreV2 using only judgements on the results screen, or at least the worst case scenario. Below are changes that can be made to get as close as possible to improving the results screen score formula.

The changes:

I will preface that these changes are for osu! only. However, these concepts may apply to the other rulesets as well.

Here's a Desmos with all the improvements implemented: https://www.desmos.com/calculator/ggriho3ymj

For the accuracy multiplier:

- Add an exponent, in this case, `acc^10`, which is what ScoreV2 uses. In the old lazer score formula, 95% accuracy is treated as a 0.95x multiplier on the 300k score dedicated to accuracy, which is only 15k score taken off from the total. This is a lower penalty compared to missing at 280/300 combo, which will net you 93% of the 700k dedicated to combo, equalling a 53k penalty. This turns the leaderboards from a mixture of accuracy and combo skill to just combo game, frustrating players when they improve their accuracy, but obtain lower score due to, for example, missing earlier. Since the 75% to 70% accuracy portion penalty should be less harsh than the 99% to 94% accuracy penalty, an exponent works well in our favor.

For the combo multiplier:

- The combo scaling formula is [this formula](https://i.imgur.com/3ozsJQ2.png). This is what the current ScoreV2 combo scaling formula simplifies to, with the assumption that: the judgement values stay constant, and you normalize the combo value to the range of (0, 1).

- The first combo score portion, `c_1`, is just the achieved combo and the maximum combo fed into the above formula. This is a near perfect approximation of the score gained from the highest combo streak.

- Before we calculate the second combo portion, we first get the effective misscount of the play, to account for potential sliderbreaks. This formula is the same one used in the performance points system, and can be fractional.

- The second combo score portion, `c_2`, is more complicated. To start, we assume the maximum combo was set at the very start of the map, and the non-max combo portions are all one long streak. This means the non-max combo portion would be `max_possible_combo - max_achieved_combo` long. We then divide this non-max combo portion by the effective misscount, to get the length of each non-max combo streak, assuming every miss is evenly spread out. We feed this value into the combo scaling formula, along with the total combo of the map, to get the combo score achieved for each combo, and we multiply it by the effective misscount to account for the number of non-max combos. This value is the worst case scenario for non-max combo scaling.

- Afterward, we add these together to get the semi-final combo portion value.

- To finalize, we multiply this value by the raw accuracy value of the score. This is the average loss of combo score from 100s and 50s hit during the play.

For score progression:

Score progression is quite bad when you implement these changes on their own, so here are some improvements you can make to keep it as close to true ScoreV2 as possible:

- Keep accuracy progression, and use the current accuracy value instead of the minimum achievable accuracy value which was previously in lazer score.

- `c_1` can remain the same, but `c_2` needs one small change. When getting the length of the non-max combo portion, the maximum possible combo should be the total combo of every object currently played through, not the total combo of the entire map. This makes no difference in the end result, but keeps the score from being inflated early in the map.

A video showcasing score progression in comparison to true ScoreV2 can be seen [here](https://youtu.be/2UidS4lE5lQ).

**Comment by peppy** [comment]
I've read through this twice and don't really understand what you are attempting to convey.

- Could you remove everything about "scores being recalculable"? It's probably taking away from what you are proposing.
- As far as I can tell, you're saying we should revert new standardised scoring and instead make changes to old standardised to bring it closer to what a user may expect. But I'm not sure.
- Can you attach a diff showing the changes applied, or open a draft PR? It looks like you have it implemented from the attached video.

**Reply by Natelytle** [reply]
- Noted, I passed it along to Rekunan, but it'll have to wait for when he wakes up.
- Yep, that's what I'm trying to go for. The compromise of raw ScoreV2 being the path forward but this method being employed for all legacy and old lazer scores is something I would also be happy with.
- It's very crudely implemented as it's my first ever time messing around with score, and it's in an old master branch where results screen standardized scoring was already implemented. [Here](https://github.com/Natelytle/osu/tree/results-scorev2)'s a branch with the changes. They are currently not completely accurate on leaderboards which has to do with legacy combo increase not working unless the replay is viewed?? You can fix it by replacing `current.TotalCombo` with `maximum.MaxCombo` in all occurrences, but it ruins score progression.

**Comment by WitherFlower** [comment]
> He has the standpoint that score should provide information not on the results screen, such as punishing spread out misses more than grouped misses, rewarding longer non-max combos, etc. This is all completely subjective and is not how I view score.

I suggest you read the following comment, or at least the combo-related paragraph for more reasons as to why this direction was preferred over "result screen scoring". https://github.com/ppy/osu/discussions/20717#discussioncomment-4071089

Not considering the length of non-max combos is not a change you can justify by saying "this is not how I view score". It has a big impact on game balance, and I think it warrants a more thorough explanation. 

Furthermore, plays coming from stable have chain-misses more often than evenly spread out misses due to stable having strong notelock, making your assumption of spread-out misses suboptimal.

**Reply by Natelytle** [reply]
I'll think of a way to put my feelings towards the current direction for score into better words, and I'll get back to you on that point when I do. However, I don't think your assumption that chain misses are more common is accurate. I only see it on speed or spaced stream plays, which are common in the higher levels, but the majority of the player base plays jump maps in which chain misses are more uncommon than spread out ones. 

Addressing the first point (the section about chain misses in particular), you can make a case that on jump and burst maps, chain misses show a greater lack of skill than a few spread out ones, as missing on a burst part or missing many notes in a jump section in a row show a greater lack of skill on those skill sets than missing on easier sections due to nerves or just being unlucky. In an effort to not make score another variant of PP, I believe giving an equal penalty to all misses is a fair compromise.

Many of the points in the write-up you linked make sense when looking at them through the lens of one type of beatmap, but fall apart when you apply them to the game as a whole. For example: 

> A 300 in a long high accuracy section is more impressive than a 300 alone, and from the other viewpoint one “big” timing mistake (late at stream start, lacking stamina, did not expect 1/3rd section, etc… causing multiple 100s/50s) has no reason to cost you that much more score than other timing mistakes.

This point applies very well to stream maps, I must admit. But in a map like Sputnik-3 or Decaplets where getting low accuracy on the finger control parts shows a greater lack of skill than getting a few 100s dotted across the map, this point is counterintuitive and the leaderboards for those maps would reflect that.

**Comment by Zyfarok** [comment]
> people dislike ScoreV1 for putting plays with worse max combo, acc, misscount, etc

I agree for the acc/misscount part but not for maxcombo, and for most, the reason is NOT because they want score to be 100% determined from those numbers but because scorev1 is bad in a lot of ways: Mostly underweighted accuracy, but also different accuracy weight between start and finish of the map/combo, low impact of lower combos, and underweighted mod multiplers (mostly DT).

We could argue that it might however be useful to store and display more information for explainability of the score, like top 5 combos (and combo-score, average combo, etc...), even if the scoring formula does not restrict itself to them. However reducing the fairness of scoring for people that only look at 3 numbers on the score screen seems to be the worst idea.

> The only reason ScoreV1 can be converted to ScoreV2 so accurately at the moment is because the combo portion of ScoreV2 is very similar to ScoreV1. However, if for any reason the combo scaling formula is changed for balancing reasons, this accuracy would decrease, potentially to a level undesirable for leaderboards.

First, Indeed the changes I suggested to make over scorev2 (mostly adding an exponent ~0.5 on combo) does make the conversion slightly worse. However, the conversion can still be very good. In fact, by reverse-engineering the v1 formula it is possible to evaluate a "worst-case scenario" (and best-case if desired) for the distribution of combos/misses better than if using only acc/misscount, which allows to compute a score which wouldn't be too far off (except for maps with 0 combo scaling like you mentioned).
Leaderboards of easier maps are filled with FCs and SSs which are trivial to convert "perfectly" (with the exception of slider-acc), so I suppose your concern is for harder maps with no FCs, but the difference of scores between the 1st player and the others on those maps are generally big enough that approximation should not affect the 1st place. And the closer we are to an FC, the better the conversion can be !
I can't do it now but I can provide some numbers next week if desired.

Also, those non-FC leaderboards usually continue to evolve with time as the player-base keeps improving and thus prioritizing the future of scoring over the accuracy of conversion of old score seems be a much better choice to me.

**Comment by smoogipoo** [comment]
Just to make sure I understand correctly... is the main point of this that we can approximate SV2 with the old standardised scoring, and we should instead look for improvements via that route so that scores can continue to be re-calculable?

Primarily, the reason for using ScoreV2 is that we're looking to get lazer "released" and scores ratified so they can be displayed in profiles/given PP/affect global rankings/etc. It's one less hurdle for us to cross if we're able to use something that has generally been proven to work over almost a decade, which the old standardised scoring hasn't been able to do.  
As I said before, and will reiterate, I'm perfectly okay with changing scoring to use something more akin to old standardised and keeping scores re-calculable, as long as the method has been proven. As for how to prove it? That's going to take more than a discussion thread.

Regardless, if you maintain that you can keep scores re-calculable from the results screen, then we should be able to apply these changes at any point within lazer's lifespan, and so this can be a continuous work-in-progress that we don't need to tend to right away, correct?

> There are ScoreV1 edge cases that need accounting for. As an example, the bottom diff of [this mapset](https://osu.ppy.sh/beatmapsets/325158#osu/741477) has a multiplier of 0 on the combo scaling portion, which means every judgement is worth the same amount. This causes a linear score progression curve, as opposed to a quadratic one. This is a problem, since if the score is converted using the proposed methods, scores with the same combo value set in lazer will be way less than their corresponding stable scores.

The point of the conversion isn't to make it so that same combo ~= same score, in-fact this is not going to be the case even outside of this extreme case since we've split accuracy up into its dedicated 30% portion - the conversion is adding in another "accuracy scaling" portion that is unique to sv2.
The point of the conversion is first to bring scores into the 1mil range thereby keeping the raw score value achievable from a play made in lazer, and then to roughly maintain leaderboard order while applying said scaling.

**Reply by Zyfarok** [reply]
> It's one less hurdle for us to cross if we're able to use something that has generally been proven to work over almost a decade

Scorev2 has generally been proven to be much better than scorev1 but still unfair in many scenarios.

> Primarily, the reason for using ScoreV2 is that we're looking to get lazer "released" and scores ratified so they can be displayed in profiles/given PP/affect global rankings/etc.

Getting lazer "released" means the scores made on lazer after that will have to be kept, and thus some changes to the formula will be impossible to make unless you accept those lazer scores to be converted approximately again, like current v1 scores.
I hope you're taking that into account and not planning to rush the scoring formula to then lock most aspects of the formula as soon as lazer is released.

**Reply by integerrr** [reply]
Nothing will have to be "converted" if a re-calculable score formula is implemented down the road as opposed to making changes to the sv2 formula which is what you're implying if I haven't misunderstood you. 

From what I gather this is what's been happening:
- the previous lazer standardised scoring doesn't work great and further work is required
- this is work which requires time from devs
- which is time that can be better used on getting the rest of the game ready/to feature parity
- sv2 in comparison works pretty great in most scenarios already with good user acceptance, takes minimal time to get working

So no, whatever scoring changes doesn't _necessarily_ mean we will only be able to work on the sv2 formula

**Reply by Zyfarok** [reply]
A "re-calculable" formula based solely on the values currently stored in the DB will necessarily have issues, this has been discussed before (check https://github.com/ppy/osu/discussions/20717#discussioncomment-4071089), thus the need to either go for something non-recalculable or storing more values into the database. Scorev2 itself is a base that would be better than anything recalculable, but it's not recalculable and hard to improve without more data.

**Comment by Zyfarok** [comment]
I've come to think of a compromise that I believe could help satisfy both the "recaculable" and the "non-recalculable" sides of the discussion.

The main problem to make a formula recalculable is that not enough data is stored in the database to create a fair scoring formula encompassing all combos in a map. 
On the other hand, the argument for a recalculable formula is both to allow easier tweaks of the formula in the future AND also allow for players to have a better understanding of how the score is computed and why X is a higher score than Y.

There has been multiple suggested solutions : Storing the top X combo, storing all combos or all judgements, storing a "ComboScore" separately, etc.
All those solution have various issues (leading to bad formula or leading to storage and "understandability" issues).

Thus, from what I understand, the solution is probably to store a few more numbers that can easily be interpreted by players, and that can be used to make a better recalculable formula.

One such number is the "(weighted) average combo":  
Basically the sum of squared combos, divided by the maximum achievable combo on the map. On a given map, this number is roughly proportional to the ComboScore part of score v2 scores or even score v1 (without the acc contribution), but simplified and in a format that could be displayed on the score screen and understood by players. It can be estimated from a v1 score with great accuracy (except for edge-case maps) similarly as was discussed before.

And my suggestion is to also add a "(weighted) average square-root combo":
Basically the sum of combos to power 1.5 (instead of 2 above), divided by the maximum achievable combo on the map. While this number is a bit harder to interpret by players, I believe it is still relatively easy to describe, especially since it is a small tweak of the "average combo" above. This value can also be estimated from a v1 score with a relatively good accuracy, even if the accuracy of the conversion will be slightly lower on high miss-count scores. (I won't describe how in this message, to keep it short)

While it is possible to create a good formula with the "average combo" alone, I believe we can achieve must better results with the "average square-root combo" for the reason I mentioned in the other thread (https://github.com/ppy/osu/discussions/20717#discussioncomment-4071089), and nothing stops us from going back to only using the "average combo" in the future if the community prefers that. I believe this solution would make it easier to have a formula that is recalculable while fairly rewarding all combos of scores and avoiding the pitfalls of v1 and v2 that make 1-missing a maps cost way too much score.

Here is an example score formula using "average square-root combo" (very similar to what was suggested in the above mentioned thread):
```
700_000 * (average_square_root_combo / max_average_square_root_combo) * (acc ^ 3) + 300_000 * (acc ^ 8)
```

**Reply by peppy** [reply]
Is that number any more useful than just seeing the combo / accuracy portions of the score?

aka:

![Safari 2023-06-26 at 06 01 53](https://github.com/ppy/osu/assets/191335/de29b1a9-f632-4d6d-87c1-2054f139492a)

Because this is the direction we're going on the results screen.

**Reply by Zyfarok** [reply]
I don't know, but I believe those numbers would be good to display (with %age of max value on hover ?), as they are somewhat self-explanatory (the average combo more than the average square-root combo I admit, but still) and help understand in what aspect X score is better than Y score.

The combo / accuracy portions could be displayed (probably along with the %age too) on hover only but I believe they are harder to interpret (people will not necessarily get what caused the ComboScore to be higher for X score than Y score, given than Y score has higher max-combo for example)
